test/Microsoft.DotNet.Watcher.Tools.FunctionalTests/AwaitableProcess.cs

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.Extensions.Internal;
+using Xunit.Abstractions;
+
+namespace Microsoft.DotNet.Watcher.Tools.FunctionalTests
+{
+    public class AwaitableProcess : IDisposable
+    {
+        private Process _process;
+        private readonly ProcessSpec _spec;
+        private BufferBlock<string> _source;
+        private ITestOutputHelper _logger;
+        private int _reading;
+
+        public AwaitableProcess(ProcessSpec spec, ITestOutputHelper logger)
+        {
+            _spec = spec;
+            _logger = logger;
+        }
+
+        public void Start()
+        {
+            if (_process != null)
+            {
+                throw new InvalidOperationException("Already started");
+            }
+
+            var psi = new ProcessStartInfo
+            {
+                UseShellExecute = false,
+                FileName = _spec.Executable,
+                WorkingDirectory = _spec.WorkingDirectory,
+                Arguments = ArgumentEscaper.EscapeAndConcatenateArgArrayForProcessStart(_spec.Arguments),
+                RedirectStandardOutput = true,
+                RedirectStandardError = true
+            };
+            _process = Process.Start(psi);
+
+            StartProcessingOutput(_process.StandardOutput);
+            StartProcessingOutput(_process.StandardError);;
+        }
+
+        public Task<string> GetOutputLineAsync(string message)
+            => GetOutputLineAsync(m => message == m);
+
+        public async Task<string> GetOutputLineAsync(Predicate<string> predicate)
+        {
+            while (!_source.Completion.IsCompleted)
+            {
+                while (await _source.OutputAvailableAsync())
+                {
+                    var next = await _source.ReceiveAsync();
+                    if (predicate(next))
+                    {
+                        return next;
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        private void StartProcessingOutput(StreamReader streamReader)
+        {
+            _source = _source ?? new BufferBlock<string>();
```

Comment:
for some reason I feel uneasy with `BufferBlock` - feels like something from a different world...